arrows/vxl/color_commonality_filter.cxx

```diff
@@ -0,0 +1,561 @@
+// An optimized (unsafe if used incorrectly) function which populates
+// a n^p dimensional histogram from integer image 'input' given the
+// resolution of each channel of the histogram, and a bitshift value
+// which maps each value to its channel step for the histogram
+template < class InputType >
+void
+populate_image_histogram(
+  vil_image_view< InputType > const& input, unsigned* hist_top_left,
+  unsigned const bitshift, std::ptrdiff_t const* hist_steps )
+{
+  // Get image properties
+  auto const ni = input.ni();
+  auto const nj = input.nj();
+  auto const np = input.nplanes();
+  auto const istep = input.istep();
+  auto const jstep = input.jstep();
+  auto const pstep = input.planestep();
+
+  // Filter image
+  InputType const* row = input.top_left_ptr();
+  for( unsigned j = 0; j < nj; ++j, row += jstep )
+  {
+    InputType const* pixel = row;
+    for( unsigned i = 0; i < ni; ++i, pixel += istep )
+    {
+      std::ptrdiff_t step = 0;
+      InputType const* plane = pixel;
+      for( unsigned p = 0; p < np; ++p, plane += pstep )
+      {
+        step += hist_steps[ p ] * ( *plane >> bitshift );
+      }
+      ++( *( hist_top_left + step ) );
+    }
+  }
+}
+
+// ----------------------------------------------------------------------------
+// Integer-typed filtering main loop
+template < class InputType, class OutputType >
+typename std::enable_if< std::is_integral< OutputType >::value >::type
+filter_color_image( vil_image_view< InputType > const& input,
+                    vil_image_view< OutputType >& output,
+                    std::vector< unsigned >& histogram,
+                    color_commonality_filter_settings const& options )
+{
+  assert( input.ni() == output.ni() && input.nj() == output.nj() );
+  assert( is_power_of_two( options.resolution_per_channel ) );
+
+  if( input.ni() == 0 || input.nj() == 0 )
+  {
+    return;
+  }
+
+  // Configure output scaling settings based on the output type and user
+  // settings
+  InputType const input_type_max = std::numeric_limits< InputType >::max();
+  OutputType const output_type_max = std::numeric_limits< OutputType >::max();
+  auto const histogram_threshold = static_cast< unsigned >( output_type_max );
+  unsigned histogram_scale_factor = options.output_scale_factor;
+
+  // Use type default options if no scale factor specified
+  if( histogram_scale_factor == 0 )
+  {
+    histogram_scale_factor = histogram_threshold;
+  }
+
+  // Populate histogram steps for each channel of the hist
+  std::vector< std::ptrdiff_t > histsteps( input.nplanes() );
+  histsteps[ 0 ] = 1;
+
+  for( unsigned p = 1; p < input.nplanes(); ++p )
+  {
+    histsteps[ p ] = histsteps[ p - 1 ] * options.resolution_per_channel;
+  }
+
+  // Fill in histogram of the input image
+  unsigned const bitshift =
+    integer_log2( ( static_cast< unsigned >( input_type_max ) + 1 ) /
+                  options.resolution_per_channel );
```

Comment:
I chose to keep the input specified as a value and not in bits, as the raw value is used frequently. However, I updated the documentation and error checking.

Reply:
> as the raw value is used frequently

Okay, so to "can we change the specification of resolution to be bits rather than values?", that'd be a "no" :slightly_smiling_face:. That's fine; was just a thought.